### PR TITLE
feat(lua): API for overriding speaker volume

### PR DESCRIPTION
**Pull request by mha1**
new LUA API functions:
- audioQueueIsEmpty(): returns true if audio queue is empty 
- volumeSet(value): allows LUA to override radio settings, GF volume and SF volume. Value in [0..23] (silence to max volume)
- volumeSetClear(): clears the LUA volume override set by volumeSet(value). Volume returns to radio settings, GF volume and SF volume 

Example use case:
- system volume is set low
- LUA script wants to issue warning or important values at higher volume

Example pseudo code:
```
local function run()
  if criticalWarning then
    volumeSet(20)
    playTrack("<path to wav>")
    playTone(0, 0, 10)          -- add short period of silence to avoid audio queue empty before track finishes
  end

  if audioQueueIsEmpty()then
    volumeSetClear()
  end
end
```

### CODE REVIEW

**Comment by 3djc** [comment]
I'm not too sure why lua should have the power to override user's settings or SF. It needs at the very least a way to prevent that override

**Comment by mha1** [comment]
See my example use case. And there is no override if you don't use volumeSet().

**Comment by 3djc** [comment]
If don't want some BF script or you name it other script to screen out loud at night because some lua dev decided that this announcement must be loud when I have set my volume low or muted by whatever means

**Comment by mha1** [comment]
you are not running random scripts, are you?

**Comment by 3djc** [comment]
The super vast majority of users are running scripts written by others...

**Comment by mha1** [comment]
here's one of my use cases. flying my glider with fellow pilots. I have a script running that checks some vitals, eg rx battery voltage reporting threshold violations. to not annoy my fellow pilots with my vario beeping I turn down the volume to the extent of nearly hearing it and frequently not hearing it when my fellow pilots are chatting. no problem for vario but if a warning hits I want it at let's say medium volume. as far as i know - and i asked in the lua channel without luck - there's no way to achieve this right now.

**Comment by 3djc** [comment]
I'm unsure why your warning would require lua, but isn't vario on the specific beep volume level ?

**Comment by mha1** [comment]
It's not just vario but also on demand eg altitude readings, flight time etc.

And it is very convenient to have all warning stuff in a single lua script rather than umpteen logical variables and play tracks. Might not be your use cases but imho still valid.

**Comment by JimB40** [comment]
Michael,

I'm in the process of review of luadoc. 
Since part of API description is in a code there will be a guide line to make proper one.
It will take some time but I will monitor new functions so they have proper luadoc info.

So.
1. Func naming
Convention (90%) is to start with verb
a. volumeSet(value) -> setVolumeOverride(value)
b. volumeSetClear() -> deleteVolumeOverride()
c. audioQueueIsEmpty() -> isAudioQueueEmpty()

BTW last one this seems to be very specyfic function, maybe w should go with more generic one like getAudioQueue()
if there are no sounds queued returns nil or 0 otherwise returns number of queued sounds.

2. Parameters.
a. type 
param volume (number) -> param volume (integer)
b. range
(0..VOLUME_LEVEL_MAX) -> (0..23)
following by info that  23 is defined as VOLUME_LEVEL_MAX constant
 
3. Return value(s)
If function is not returnin value please add
@retval none

4. Targets info
We need info if function works on BW or COLOR only or both.
I'll be talking with @pfeerick about parser that generates luadoc, but for now simple comment
BW, COLOR
is  enough

4. Last but not least we need some a way to add searchable tags in code for later luadoc update
My proposition is to add block of comments below /*luadoc section

In this case it should look like:
// LUADOC TODO:
//  - Add  VOLUME_LEVEL_MAX contant to CONSTANTS section
//  EXAMPLES:
// https://github.com/EdgeTX/edgetx/pull/3784

Once we update luadoc, those comments will be deleted

IDK if this is valid way, if not - propse something else that can be created by dev at the time of PR merge. 
There are many API funcions without proper API info, and onlu author knows how it exacly works while not eveybody of lua  devs is interested in reading C code to find out.

**Comment by gagarinlg** [comment]
I like the idea of LUA being able to override the volume in certain cases.
Maybe we could add an option in the radio settings to limit the LUA volume to the global volume of this is desired and make that the default. When someone specifically wants to allow Lua to be louder then the system volume that person has to allow it.

**Comment by mha1** [comment]
@JimB40 Robert, thanks for your suggestions, will update the PR.

isAudioQueueEmpty() has its root in class AudioQueue which just returns a boolean. For my use case it is sufficient to now if an issued warning is still playing or not to determine when to cancel the override.

Any thoughts on proceed or scrap this PR? JC is obviously not a fan and I can understand his slightly pessimistic point of view.

**Comment by JimB40** [comment]
> When someone specifically wants to allow Lua to be louder then the system volume that person has to allow it.

I undestand @3djc concern, but there are many func that can annoy user if used not properly. Good devs will make it as an option. Bad ones ... well people do not use bad soft :)

**Comment by mha1** [comment]
Yes, have faith

**Comment by JimB40** [comment]
> isAudioQueueEmpty() has its root in class AudioQueue which just returns a boolean. For my use case it is sufficient to now if an issued warning is still playing or not to determine when to cancel the override.

AFAIR every new function declatation cosumes memory (BW problem mainly) so approach that one function can handle more than one feature is more memory-frendly.
To my experience once func is added it is quite hard to depreciate it when used in many scripts. That's why I'd prefer something more like AudioQueue status, that can later be extended if needed.

**Comment by mha1** [comment]
@JimB40 fair enough. will get back to you with a proposal how to combine set/clear into one call and returning isEmpty as return parameter audio status

**Comment by mha1** [comment]
@gagarinlg I wouldn't opt for a user setting. Put in perspective this might be a nuisance at best if misused. You don't have settings for allowing messing up screens, overriding channels, creating and remote controlling SF's, etc you name it.

**Comment by 3djc** [comment]
We already have requests to make volume SF kicks before any sound gets emitted and overrides default sound level. And this proposes to simply ignore it without any user way to restrict it. Contradictory imho.

if this is to be implemented, it needs at least a way for the user to protect against it

**Comment by gagarinlg** [comment]
I personally do not care if there is an option to limit Lua audio levels to the system level or not. It just seems to be a good compromise to have the option. If it is opt-in or opt-out is another discussion.

**Comment by raphaelcoeffic** [comment]
@3djc we even had the plan to remove the GF and use a directly assignable control, so that the LUA function would be an alternative for those who still want something like that.

That being said, I understand your concerns as well. Not sure if the solution is "yet another setting" (we have already so many options!).

@mha1 for my part, I'm a bit concerned that we cannot ensure that the override is gone once the script is gone. So a script could be stopped or removed, leaving the user with no control at all over the volume. Only a restart would fix this. Not sure we can find a solution to this though (automatic cleanup?).

**Comment by raphaelcoeffic** [comment]
Thinking a little longer about it, we could use a decaying timer value instead of a boolean for the override, so that the LUA script would be forced to set the volume for example at least every second, otherwise the override would gone, which would solve the issue I presented in my previous comment.

**Comment by 3djc** [comment]
> @3djc we even had the plan to remove the GF and use a directly assignable control, so that the LUA function would be an alternative for those who still want something like that.

I did read correspondiing thread, and it was clearly mentionned there that SF would NOT be removed by that (https://github.com/EdgeTX/edgetx/pull/2256#issuecomment-1623010406) , as having a single control on audio does not meet the average use case

**Comment by mha1** [comment]
@raphaelcoeffic I like the cleanup suggestion as it would probably make sense for other things too. 

Maybe luaDisable() and luaError() would be a good starting point. luaDisble() already does a tiny bit of cleanup and would as it seems catch a LUA shutdown. Failing scripts would go through luaError(). Although not script specific luaError() could reset some general stuff like a volume override.

But in general this PR is low emotional debt to me. If Y'all think a LUA function shouldn't mess with volume let me know and I'll withdraw. So before discussing solutions I'd like to have a vote on the question "Should LUA scripts be allowed to override volume settings?". To advocate my Yay I'll just say you can do much more damage with the existing functions.

**Comment by JimB40** [comment]
Volume and LCD brightnes are very often used and with current implementation via GF/SF feature have twisted logic.

Should be:
First: What to do
Second: how to do

In example going to Backlight I have:
- ON brightness - two options Value ( UI system sliders) and Source (whatever: CH, LS, Analog etc)
- OFF brightness - two options Value ( UI system sliders) and Source (whatever: CH, LS, Analog etc)

**Comment by raphaelcoeffic** [comment]
Something we just discussed with @3djc: we should probably first clarify the use-case.

The idea was the following:
- either you have a dedicated control (see #2256; basically replacing the SF with a knob or trim-like buttons),
- or you can implement it in LUA if you prefer.

That would limit the use to advanced volume control, rather than allowing some app devs to just push the volume up to issue annoying warnings.

@mha1 what is actually you original use-case?

**Comment by pfeerick** [comment]
Probably need to go further back, and redesign this whole thing from the start, as it is a convoluted mess atm. i.e. people complete that GF controlling volume doesn't work during pre-flight checks. If the SF/GF is capability is removed, I know that people who use them (myself included) would complain. etc, etc, etc. There may become a point where we won't have a happy median, but lets see if we can get there.

So, taking into account https://github.com/EdgeTX/edgetx/pull/2256 and https://github.com/EdgeTX/edgetx/pull/3784... and with both backlight/brightness AND volume in mind... 

- Doing away with the GF, I don't see that being a problem, if it's now a system-level setting, which can also potentially be a bit smarter. caveat: Except, you do have the ability to disable able GFs on a per-model basis.... just saying it's a possibility... 
- I think having the SF is still needed for per-model control, unless some other mechanism is come up with that somehow makes sense. Mainly because the logic is there so that you can control stuff however and whenever you want. 
- Then, Lua. Yes, I can see that being a problem (and also something that is _missing_). Lua shouldn't be able to bypass the sound setting. i.e. if I have it set to 'Quiet', Lua doesn't have a say in the matter, it's not aloud to mess with my sound settings, or make a peep. 

So, the question I then see is, how/do we need to give the user options as to whether Lua can change the volume / make sound? Or is this something that the app developers would need their users to give them an earful about since they didn't offer quiet or sound options in their app? In other words, this isn't our problem, and we're overthinking this?

**Comment by raphaelcoeffic** [comment]
@mha1 thinking about you use-case, there is something that we may be missing here: sound is played asynchronously. This means that changing the general volume for playing certain sounds is not going to work very easily.

I have the feeling that what you actually want is a different volume based on the *type* of sound, meaning that we need to have more categories and relative volume setting than we have now.

Currently, we have this:
- "Volume" (basically the baseline for maximum volume)
- "Beep Volume" (relative to "Volume)
- "Wav Volume" (relative to "Volume")
- "Background Volume" (relative to "Volume")
- "Variometer Volume" (relative to "Volume")

And it seems you would like some "Warning Volume" (to be defined what is really is). So basically another relative volume for things you consider need to be played louder than the "Variometer" and possibly different announcements.

**Comment by pfeerick** [comment]
Or is it more "Lua volume"? The thought that the volume that a Lua script can play a sound could be different to the rest of the system ("Wav Volume"/ "Beep Volume")

**Comment by raphaelcoeffic** [comment]
> Or is it more "Lua volume"? The thought that the volume that a Lua script can play a sound could be different to the rest of the system ("Wav Volume"/ "Beep Volume")

I need to check, but maybe we can include the *relative volume* to the audio queue, so that the script can choose on a case-by-case basis. That being said, the volume would not be able to be more than what chose (which seems important to most folks).

**Comment by raphaelcoeffic** [comment]
Ok, just discussed it with @3djc: the idea is to add a `volume` attribute to the `AudioFragment`, that would allow to override the volume used by the `mixBuffer()` functions of contexts.

```c++
struct AudioFragment {
  uint8_t type;
  uint8_t id;
  uint8_t repeat;
  // TODO: add 'volume' here
  union {
    Tone tone;
    char file[AUDIO_FILENAME_MAXLEN+1];
  };
};
```

The good news is: because of the alignment requirements, there is already a *hole* of 1 byte between `repeat` and `union {...}`, so that it would not even cost us anything.

See bellow for a fair idea of how the audio mixer works:
Screenshot 2023-07-11 at 10 00 57

Please note that each "context" has a `AudioFragment` as an attribute, so that the `volume` attribute should be available in the contexts as well when `mixBuffer(...)` is called.

The LUA code could then call `playFile()` or `playTone()` with whatever relative volume it might want to use as a parameter. All we need is to define some invalid value for this volume, so that in that case the "normal" volume (depends on context) would be used.

This means the current PR should be modified so that:
- the general volume is not touched,
- `AudioFragment` has an additional attribute `volume`,
- `mixBuffer()` methods make use of that `volume` attribute (if not invalid),
- LUA `playFile()` / `playTone()` has an additional optional `volume` parameter.

@mha1 I think this proposal should cover your use-case, as well as some additional use-cases. It will allow LUA scripts to issue sounds with different volume intensities without compromising the general volume. This allows the user to stay in control of the general volume while enabling many advanced use-cases.

**Comment by 3djc** [comment]
> 
> nice image, what SW?

https://excalidraw.com/

**Comment by mha1** [comment]
Implementation is now covering all acoustic LUA play functions and uses an optional parameter for overriding the volume on a by-call  basis. The general volume level is still governed by the set general volume in radio settings. The by-call volume acts in the same way as radio settings Wav and Beep volume being relative the general volume. playFile(), playNumber() and playDuration() temporarily replace Wav volume, playTone() temporarily replaces Beep volume.

Simple test cast 1:
- set radio volume slider to lowest level (pos 3 of 5)
- set Wav volume to lowest level (pos 1 of 5)
- set up LUA script to issue:
playFile("armed.wav",2)  -- play file armed.wav, volume 2 -> will sound like Wav volume set to highest level (pos 5 of 5)
playFile("armed.wav",-1) -- play file armed.wav, volume -1 -> will sound like Wav volume set to low level (pos 2 of 5)
playFile("armed.wav", 0) -- play file armed.wav, volume 0 ->  will sound like Wav volume set to mid level (pos 3 of 5)
playFile("armed.wav")     -- play file armed.wav, volume omitted ->  will sound according to radio settings Wav volume (in this test case lowest, pos 1 of 5)

Test case 2:
-change radio settings general Volume to high level (pos 4 of 5)
- run script again and note audio output at generally higher level and playFile("armed.wav")  playing at high level

**Comment by JimB40** [comment]
@pfeerick I will monitor PRs changing LUA API, but to double check if you spot new one without in-code proper API luadoc comments (including examples) like @mha1 did here, pls do not merge until updated.

**Comment by mha1** [comment]
@pfeerick @raphaelcoeffic @3djc please review

File: radio/src/audio.h [hunk 1]
```diff
@@ -156,31 +158,36 @@ struct AudioFragment {
   uint8_t type;
   uint8_t id;
   uint8_t repeat;
+  int8_t _volume;
```
Comment (raphaelcoeffic) [inline radio/src/audio.h:161]: Any particular reason for using `_volume` instead of `volume`?

File: radio/src/audio.h [hunk 2]
```diff
@@ -156,31 +158,36 @@ struct AudioFragment {
...
   union {
     Tone tone;
     char file[AUDIO_FILENAME_MAXLEN+1];
   };
 
   AudioFragment() { clear(); };
 
-  AudioFragment(uint16_t freq, uint16_t duration, uint16_t pause, uint8_t repeat, int8_t freqIncr, bool reset, uint8_t id=0):
+  AudioFragment(uint16_t freq, uint16_t duration, uint16_t pause, uint8_t repeat, int8_t freqIncr, bool reset, int8_t _volume, uint8_t id=0 ):
```
Comment (raphaelcoeffic) [inline radio/src/audio.h:169]: `volume`

File: radio/src/audio.h [hunk 3]
```diff
@@ -156,31 +158,36 @@ struct AudioFragment {
...
     type(FRAGMENT_TONE),
     id(id),
     repeat(repeat),
+    _volume(_volume),
     tone(freq, duration, pause, freqIncr, reset)
   {};
 
-  AudioFragment(const char * filename, uint8_t repeat, uint8_t id=0):
+  AudioFragment(const char * filename, uint8_t repeat, int8_t _volume, uint8_t id = 0):
```
Comment (raphaelcoeffic) [inline radio/src/audio.h:177]: `volume`

File: radio/src/lua/api_general.cpp [hunk 4]
```diff
@@ -1460,16 +1476,32 @@ Play a numerical value (text to speech)
...
+@retval none 
+
+@status current Introduced in 2.0.0, changed in 2.10
 
+// targets: BW, COLOR
+//
+// EXAMPLES:
+// playNumber(123, 3, 0, 2)  -- play number 123, unit mAh, use Wav volume 2
+// playNumber(123, 3, 0, -1) -- play number 123, unit mAh, use Wav volume -1
+// playNumber(123, 3, 0)     -- play number 123, unit mAh, use Wav volume 0
```
Comment (raphaelcoeffic) [inline radio/src/lua/api_general.cpp:1493]: Isn't `playNumber(123, 3, 0)` using default volume?

File: radio/src/audio.h (hunk 1 above)
Comment (mha1) [reply radio/src/audio.h:161]: This was meant to make distinguishing the existing volume from the overriding volume. Finally _volume gets calculated into volume.

File: radio/src/audio.h (hunk 2 above)
Comment (mha1) [reply radio/src/audio.h:169]: same here.

File: radio/src/audio.h (hunk 3 above)
Comment (mha1) [reply radio/src/audio.h:177]: same here

File: radio/src/lua/api_general.cpp (hunk 4 above)
Comment (mha1) [reply radio/src/lua/api_general.cpp:1493]: No. Consider these radio settings:
image
They are the "user default" settings (don't confuse this with the system default settings). 

playNumber(123,3,**0**) will effectively override the Wav volume setting for the playNumber duration to
image
which is different to the "user default" and different to the "system default" (as user changed overall volume too).

**Comment by 3djc** [comment]
I think the choosen scale is very misleading then: - 0 + suggest a relative volume, and what you are saying now is that it is an absolute one. It should be 1 to 5 then

File: radio/src/lua/api_general.cpp (hunk 4 above)
Comment (pfeerick) [reply radio/src/lua/api_general.cpp:1493]: But isn't volume the **4th** parameter? 
> playNumber(value, unit [, attributes [, volume]]

Comment (mha1) [reply radio/src/lua/api_general.cpp:1493]: @raphaelcoeffic @pfeerick I shouldn't respond in a hurry. Of course, omitting the last parameter doesn't override any of the user settings, ie the user radio settings Wav volume is applied. This also how compatibility is ensured.

**Comment by mha1** [comment]
-2...2 is the internal representation of the Wav and Beep volume slider and I think they are actually a relative volume. 0 means unadjusted general volume. A negative value makes wav's or beep's lower than general volume, a positive value higher.

File: radio/src/lua/api_general.cpp (hunk 4 above)
Comment (mha1) [reply radio/src/lua/api_general.cpp:1493]: omitting the forth parameter is identical to setting the forth parameter to USE_SETTINGS_VOLUME. Omitting it is for backwards compatibility

**Comment by 3djc** [comment]
> -2...2 is the internal representation of the Wav and Beep volume slider and I think they are actually a relative volume. 0 means unadjusted general volume. A negative value makes wav's or beep's lower than general volume, a positive value higher.

I’m sorry I don’t understand. It comes default with all value maxed. What do you call general volume?

**Comment by mha1** [comment]
A quick test should reveal. Set general volume to mid level, set wav volume to mid level. Have a SF to play a track repeatedly, eg. 3s. go to radio setting listen to the track played at wav volume mid. change the wav volume slider to max wav volume. The track should play louder. Leave wav volume at max and reduce general volume. the track is now playing at lower volume. reduce wav volume to mid level. the track should now play at even lower volume.

my interpretation is the general volume sets a base level, wav and beep volume adjust wav's and beep's relative to the base level up and down with the mid level meaning (= LUA 0) no adjustment.

**Comment by mha1** [comment]
slider "Volume" just below Mode

**Comment by mha1** [comment]
Are you sure about default all at max? I don't have a radio close atm. If that's the case we are both right. wav and beep volume are relative to general volume but adjustable only in one direction. I was wondering why the internal representation is -2..2. In any case changing the scale is very easy.

**Comment by mha1** [comment]
Here's a screenshot of the radio setup sound settings right after creating a blank model. All sliders are set at mid level except `Background volume`. With general volume I mean Slider `Volume` which sets the volume base level. Note: Maybe having `Volume `not exactly centered might be a bug. Issue?
image

`Beep Volume` and `Wav Volume` act relative to `Volume`:
-2: two units less than general volume level
-1: one unit less than general volume level
0: no adjustment of general volume level
+1 one unit more than general volume level
+2 two units more than general volume level 

IMHO this is also the reason for the internal representation of the slider values (-2..2) .

This can be easily tested by setting 'Volume' to max and 'Wav volume' to mid level. Use a SF to play a track repeatedly (1s or 2s interval). Listen and while listening increase `Wav volume` to max. You'll find the track is now played louder.

Of course you can scale (-2..2) to (1..5) but you'd loose the center information. I'd like to stick with the (-2..2) interpretation and was thinking of adding LUA constants, e.g. QUITER, QUITE, NORMAL, LOUD,  LOUDER.

But as always I trust EdgeTX management's expertise and judgement. Guys, just take a vote and let me know.

**Comment by 3djc** [comment]
Center doesn't mean anything at all. Again, +1 suggest a relative volume, louder than now, but if you wav is set to max, then it is actually less volume, very confusing imho. Since min position is still having sound, I think 1 to 5 makes much more sense. (and for reference, most of the 0 storage value comes from attemps to improve RLE compression for Otx EEprom stoarge)

**Comment by mha1** [comment]
decision please

**Comment by mha1** [comment]
I give in. LUA API parameter volume now 1..5.

[voltst.lua.txt](https://github.com/EdgeTX/edgetx/files/12232677/voltst.lua.txt): qand dirty LUA function script for testing. Play with radio setup Volume, Wav and Beep Volume. The first sound of every group should follow the radio settings Wav and Beep Volume. The following five sounds of every group should scale up. Overall volume is governed by radio setup Volume setting.

**Comment by mha1** [comment]
@3djc @raphaelcoeffic Guys, you still have concerns?

**Comment by mha1** [comment]
@pfeerick silence is agreement, right?

**Comment by 3djc** [comment]
I would:

1) rename _volume to volume. It doesn't had value imho
2) your doc for playFile still mention " use Wav volume -1" 

Ok with it given the two above

**Comment by mha1** [comment]
@3djc thanks for your feedback.
- _volume: I'll change it to fragmentVolume because that's really what's at the core of this PR. Allow audio fragments to carry an individual volume setting with default using the radio settings volume.
- doc: will be updated

**Comment by mha1** [comment]
@raphaelcoeffic @3djc @pfeerick are we now good on this? Please lift my rebase curse due to continuous translation changes.

**Review by raphaelcoeffic — APPROVED** [review]
LGTM now

**Comment by 3djc** [comment]
> LGTM now

Same here

**Comment by mha1** [comment]
@pfeerick this should be good to go now

**Comment by mha1** [comment]
anyone-hello

**Review by pfeerick — APPROVED** [review]
Just had a couple of typos and erroneous changes in the luadocs. I used a modified version of the test script provided in https://github.com/EdgeTX/edgetx/pull/3784#issuecomment-1660822528 to test on X9D+2019 and TX16S, working perfectly! :) 2.10 luadocs updated accordingly.

**Comment by mha1** [comment]
Thank you!

**Pull request merged**